### PR TITLE
Save Startup Actors to snapshot :clap:

### PR DESCRIPTION
**Pull request by Vatyx**
#### Description
Startup actors, preplaced actors, actors placed in the level, stably named actors - are actors which are placed in the level.

This PR introduces support for saving startup actors to the snapshot so they are received from SpatialOS. Also adds a `Rotation` schema component to store the rotation of actors so they are spawned in the correct direction. Spawning of these entities happen normally. There might need to be some extra work done to make sure their static path in the `UnrealMetadata` component matches their path locally but so far there doesn't seem to be an issue with that.

The other side of fully supporting Startup Actors is deleting them from the world on startup, which is enabled by this PR: https://github.com/improbableio/UnrealEngine/pull/19

**This PR will not go in until the Engine change has gone in.**

#### Tests
Works in DemoGym and Cascade (I think?) in Scavengers

#### Documentation
Documentation will come when we overhaul the documentation for dynamic typebindings.

#### Primary reviewers
@mironec @m-samiec

### CODE REVIEW

File: Plugins/SpatialGDK/SpatialGDKEditorToolbar/Source/SpatialGDKEditorToolbar/Private/SpatialGDKEditorGenerateSnapshot.cpp [hunk 1]
```diff
@@ -120,52 +129,181 @@ bool CreateGlobalStateManager(Worker_SnapshotOutputStream* OutputStream)
...
+	ComponentWriteAcl.Add(ROTATION_COMPONENT_ID, UnrealWorkerPermission);
+	ComponentWriteAcl.Add(ActorInfo->SingleClientComponent, UnrealWorkerPermission);
+	ComponentWriteAcl.Add(ActorInfo->MultiClientComponent, UnrealWorkerPermission);
+	ComponentWriteAcl.Add(ActorInfo->HandoverComponent, UnrealWorkerPermission);
+	// No write attribute for RPC_Client since a Startup Actor will have no owner on level start
+	ComponentWriteAcl.Add(ActorInfo->RPCComponents[RPC_Server], UnrealWorkerPermission);
+	ComponentWriteAcl.Add(ActorInfo->RPCComponents[RPC_CrossServer], UnrealWorkerPermission);
+	ComponentWriteAcl.Add(ActorInfo->RPCComponents[RPC_NetMulticast], UnrealWorkerPermission);
+
+	for (UClass* SubobjectClass : ActorInfo->SubobjectClasses)
```
Comment (m-samiec) [inline Plugins/SpatialGDK/SpatialGDKEditorToolbar/Source/SpatialGDKEditorToolbar/Private/SpatialGDKEditorGenerateSnapshot.cpp]: This code seems messy. Two different ways of iterating through actor subobjects.
Also worth consolidating the common parts of entity/component creation to some helper functions that can be accessed at runtime.

File: Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp [hunk 2]
```diff
@@ -98,6 +99,12 @@ bool USpatialActorChannel::IsSingletonEntity()
 	return NetDriver->GlobalStateManager->IsSingletonEntity(EntityId);
 }
 
+bool USpatialActorChannel::IsStablyNamedEntity()
+{
+	// Don't delete if stable named entity
+	return !NetDriver->View->GetUnrealMetadata(EntityId)->StaticPath.IsEmpty();
```
Comment (m-samiec) [inline Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp]: What if the user wants to delete a stably named actor?

File: Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp [hunk 3]
```diff
@@ -314,8 +322,7 @@ bool USpatialActorChannel::ReplicateActor()
 			}
 
 			// Calculate initial spatial position (but don't send component update) and create the entity.
```
Comment (m-samiec) [inline Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp]: Comment is out of date.

**Comment by m-samiec** [comment]
Generally looks good - few comments.

File: Plugins/SpatialGDK/SpatialGDKEditorToolbar/Source/SpatialGDKEditorToolbar/Private/SpatialGDKEditorGenerateSnapshot.cpp (hunk 1 above)
Comment (m-samiec) [reply Plugins/SpatialGDK/SpatialGDKEditorToolbar/Source/SpatialGDKEditorToolbar/Private/SpatialGDKEditorGenerateSnapshot.cpp]: Okay that's fine, but can we create a Jira ticket to track this.

File: Plugins/SpatialGDK/SpatialGDKEditorToolbar/Source/SpatialGDKEditorToolbar/Private/SpatialGDKEditorGenerateSnapshot.cpp [hunk 4]
```diff
@@ -120,52 +129,181 @@ bool CreateGlobalStateManager(Worker_SnapshotOutputStream* OutputStream)
...
+
+void CleanupNetDriverAndConnection(USpatialNetDriver* NetDriver, USpatialNetConnection* NetConnection)
+{
+	// On clean up of the NetDriver due to garbage collection, either the ServerConnection or ClientConnections need to be not nullptr.
+	// However if the ServerConnection is set on creation, using the FObjectReplicator to create the initial state of the actor,
+	// the editor will crash. Therefore we set the ServerConnection after we are done using the NetDriver.
+	NetDriver->ServerConnection = NetConnection;
+}
+
+TArray<Worker_ComponentData> CreateStartupActorData(USpatialActorChannel* Channel, AActor* Actor, USpatialTypebindingManager* TypebindingManager, USpatialNetDriver* NetDriver)
```
Comment (girayimprobable) [inline Plugins/SpatialGDK/SpatialGDKEditorToolbar/Source/SpatialGDKEditorToolbar/Private/SpatialGDKEditorGenerateSnapshot.cpp]: This function is quite involved and does not have enough comments to described why we're doing what we're doing. Please do a comment pass.

File: Plugins/SpatialGDK/SpatialGDKEditorToolbar/Source/SpatialGDKEditorToolbar/Private/SpatialGDKEditorGenerateSnapshot.cpp [hunk 5]
```diff
@@ -120,52 +129,181 @@ bool CreateGlobalStateManager(Worker_SnapshotOutputStream* OutputStream)
...
+	// Create initial state of Actor (which is the state the Actor is in before running the level)
+	TArray<Worker_ComponentData> ComponentData = DataFactory.CreateComponentDatas(Actor, InitialRepChanges, InitialHandoverChanges);
+
+	// Actor RPCs
+	for (int RPCType = 0; RPCType < RPC_Count; RPCType++)
+	{
+		ComponentData.Add(ComponentFactory::CreateEmptyComponentData(Info->RPCComponents[RPCType]));
+	}
+
+	// Process Subobjects
```
Comment (girayimprobable) [inline Plugins/SpatialGDK/SpatialGDKEditorToolbar/Source/SpatialGDKEditorToolbar/Private/SpatialGDKEditorGenerateSnapshot.cpp]: Wherever possible, let's avoid non-helpful comments and give more color to what's going on. "Process" is usually a bad go-to word in this sense.

File: Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp [hunk 6]
```diff
@@ -98,6 +99,12 @@ bool USpatialActorChannel::IsSingletonEntity()
 	return NetDriver->GlobalStateManager->IsSingletonEntity(EntityId);
 }
 
+bool USpatialActorChannel::IsStablyNamedEntity()
+{
+	// Don't delete if stable named entity
```
Comment (girayimprobable) [inline Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp]: This comment doesn't make sense as this function has no deletion logic

File: Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp [hunk 7]
```diff
@@ -230,19 +236,10 @@ void USpatialReceiver::CreateActor(Worker_EntityId EntityId)
...
-				EntityActor = SpawnNewEntity(PositionComponent, ActorClass, true);
-				bDoingDeferredSpawn = true;
-			}
-			else
-			{
-				FString FullPath = UnrealMetadataComponent->StaticPath;
-				UE_LOG(LogTemp, Log, TEXT("!!! Searching for a native static actor %s of class %s in the persistent level whilst checking out an entity."), *FullPath, *ActorClass->GetName());
-				EntityActor = FindObject<AActor>(World, *FullPath);
-			}
+			UE_LOG(LogTemp, Log, TEXT("!!! Spawning a native dynamic %s whilst checking out an entity."), *ActorClass->GetFullName());
```
Comment (girayimprobable) [inline Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp]: Drive-by: Can you make this not LogTemp?

Comment (Vatyx) [reply Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp]: I'm planning a different PR that changes all of these `LogTemp`s to have proper log categories.
